File: src/main/java/com/oasisfeng/nevo/decorators/wechat/AssetFileProvider.java
package com.oasisfeng.nevo.decorators.wechat;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.database.Cursor;
import android.net.Uri;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Objects;

/**
 * Expose asset file with content URI
 *
 * Created by Oasis on 2019-1-28.
 */
public class AssetFileProvider extends ContentProvider {

	@Nullable @Override public AssetFileDescriptor openAssetFile(@NonNull final Uri uri, @NonNull final String mode) throws FileNotFoundException {
		final String filename = uri.getLastPathSegment();
		if (filename == null) throw new FileNotFoundException();
		final Context context = Objects.requireNonNull(getContext());
		try {
			final String suffix = context.getString(R.string.repacked_asset_suffix);
			final int offset = context.getResources().getInteger(R.integer.repacked_asset_offset);
			final AssetFileDescriptor afd = Objects.requireNonNull(context).getAssets().openFd(
					! suffix.isEmpty() && filename.endsWith(suffix) ? filename + context.getString(R.string.repacked_asset_appendix) : filename);
			return new AssetFileDescriptor(afd.getParcelFileDescriptor(), afd.getStartOffset() + offset, afd.getLength() - offset);
		} catch (final IOException e) {
			Log.e(WeChatDecorator.TAG, "Error opening asset", e);
			return null;
		}
	}

	@Override public boolean onCreate() { return true; }
	@Nullable @Override public Cursor query(@NonNull final Uri uri, @Nullable final String[] projection, @Nullable final String selection, @Nullable final String[] selectionArgs, @Nullable final String sortOrder) { return null; }
	@Nullable @Override public String getType(@NonNull final Uri uri) { return null; }
	@Nullable @Override public Uri insert(@NonNull final Uri uri, @Nullable final ContentValues values) { return null; }
	@Override public int delete(@NonNull final Uri uri, @Nullable final String selection, @Nullable final String[] selectionArgs) { return 0; }
	@Override public int update(@NonNull final Uri uri, @Nullable final ContentValues values, @Nullable final String selection, @Nullable final String[] selectionArgs) { return 0; }
}

File: src/main/java/com/oasisfeng/nevo/decorators/wechat/AssetFileProvider.kt
/*
 * Copyright (C) 2015 The Nevolution Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.oasisfeng.nevo.decorators.wechat

import android.content.ContentProvider
import android.content.ContentValues
import android.content.res.AssetFileDescriptor
import android.database.Cursor
import android.net.Uri
import android.util.Log
import java.io.FileNotFoundException
import java.io.IOException

/**
 * Expose asset file with content URI
 *
 * Created by Oasis on 2019-1-28.
 */
class AssetFileProvider : ContentProvider() {

    @Throws(FileNotFoundException::class) override fun openAssetFile(uri: Uri, mode: String): AssetFileDescriptor? {
        val filename = uri.lastPathSegment ?: throw FileNotFoundException()
        val context = context!!
        return try {
            val suffix = context.getString(R.string.repacked_asset_suffix)
            val offset = context.resources.getInteger(R.integer.repacked_asset_offset)
            val afd = context.assets.openFd(
                if (suffix.isNotEmpty() && filename.endsWith(suffix)) filename + context.getString(R.string.repacked_asset_appendix) else filename)
            AssetFileDescriptor(afd.parcelFileDescriptor, afd.startOffset + offset, afd.length - offset)
        } catch (e: IOException) {
            null.also { Log.e(TAG, "Error opening asset", e) }
        }
    }

    override fun onCreate() = true

    override fun getType(uri: Uri): String? = null
    override fun query(uri: Uri, p: Array<String>?, s: String?, sa: Array<String>?, so: String?): Cursor? = null
    override fun insert(uri: Uri, values: ContentValues?): Uri? = null
    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?) = 0
    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?) = 0
}